src/LowerTriangularMatrices/lower_triangular_matrix.jl

```diff
@@ -414,8 +414,41 @@ Base.similar(L::LowerTriangularArray{T,N,ArrayType}, ::Type{T}) where {T, N, Arr
 Base.similar(L::LowerTriangularArray{T}) where T = similar(L, T)
  
 # ARITHMETIC
+# While these would work just via broadcast (and the abstractarray interface), these explicit definitions are significantly faster
+# only mul/div with scalar and addition/subtraction, others are handled via the custom broadcast
+function Base.:(*)(L::LowerTriangularArray{T}, s::Number) where T
+    sT = convert(T, s)
+    LowerTriangularArray(L.data .* sT, L.m, L.n)
+end
+
+Base.:(*)(s::Number, L::LowerTriangularArray) = L*s         # commutative
+Base.:(/)(L::LowerTriangularArray, s::Number) = L*inv(s)
+
+function Base.:(+)(
+    L1::LowerTriangularArray{T, N, ArrayType},
+    L2::LowerTriangularArray{S, N, ArrayType},
+) where {T, S, N, ArrayType}
```

Comment:
As far as I see this this wouldn't cover the situation of `@. A = B + C` which is the non-allocating version of broadcasting that we do use inside the dynamical core, `A = B + C` isn't because of the allocation, so it's speed is much less relevant.

Reply:
Good point, I can do a small benchmark test for that on Monday

Reply:
```Julia 
using SpeedyWeather, BenchmarkTools 
NF=Float64
idims=(5,)
L = randn(LowerTriangularArray{NF}, 10, 10, idims...)
L2 = randn(LowerTriangularArray{NF}, 10, 10, idims...)
A = zeros(LowerTriangularArray{NF}, 10, 10, idims...)

@benchmark L + L2
BenchmarkTools.Trial: 10000 samples with 272 evaluations.
 Range (min … max):  291.673 ns … 60.475 μs  ┊ GC (min … max):  0.00% … 98.71%
 Time  (median):     428.287 ns              ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.107 μs ±  5.384 μs  ┊ GC (mean ± σ):  57.42% ± 11.60%

  █▃                                                           ▁
  ██▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▁▁▃▁▁▁▁▁▃▁▄▁▁▁▄▄▃▃▃▄▄▆▆ █
  292 ns        Histogram: log(frequency) by time      43.4 μs <

 Memory estimate: 2.34 KiB, allocs estimate: 2.

@benchmark L .+ L2
BenchmarkTools.Trial: 10000 samples with 9 evaluations.
 Range (min … max):  2.332 μs …  1.535 ms  ┊ GC (min … max):  0.00% … 99.61%
 Time  (median):     2.515 μs              ┊ GC (median):     0.00%
 Time  (mean ± σ):   3.349 μs ± 30.675 μs  ┊ GC (mean ± σ):  20.21% ±  2.23%

  ▅▄▇██▇▄▃▂▁▁  ▁   ▁  ▁▁▂▁▂▂▂▂▁                              ▂
  ███████████▇▇██▇▆█▇██████████▇▅▅▄▅▃▅▃▅▄▅▄▅▄▆▆▅▃▄▅▅▆▅▆▄▂▄▃▄ █
  2.33 μs      Histogram: log(frequency) by time     4.94 μs <

 Memory estimate: 2.47 KiB, allocs estimate: 4.

@benchmark @. A = L + L2
BenchmarkTools.Trial: 10000 samples with 9 evaluations.
 Range (min … max):  2.191 μs …  14.428 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.211 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.315 μs ± 464.178 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▄▃   ▂ ▂ ▁ ▁      ▁       ▂                                ▁
  ████▇████▅█▄█▅█▄▇▆▄█▄█▆▁▇▄▁█▆▃▁▁▃▁▁▃▃▁▁▁▁▁▁▄▄▁▁▃▁▆▆▅▅▆▄▁▅▅▆ █
  2.19 μs      Histogram: log(frequency) by time      4.38 μs <

 Memory estimate: 160 bytes, allocs estimate: 3.

@benchmark  LowerTriangularArray{NF, 3, Array{NF}}(undef, 10, 10, idims...)
BenchmarkTools.Trial: 10000 samples with 9 evaluations.
 Range (min … max):  2.634 μs …  1.396 ms  ┊ GC (min … max):  0.00% … 99.29%
 Time  (median):     2.787 μs              ┊ GC (median):     0.00%
 Time  (mean ± σ):   3.673 μs ± 29.901 μs  ┊ GC (mean ± σ):  19.70% ±  2.43%

  ▆██▅▂▂▂▂   ▁  ▁  ▁                                         ▂
  █████████████▇█▇███▇▇▄▄▄▅▆▇▇▇▇▇▇▇▇▆▃▅▅▃▄▅▄▃▅▃▃▃▄▃▁▁▄▄▅▆▆▆▆ █
  2.63 μs      Histogram: log(frequency) by time     6.78 μs <

 Memory estimate: 3.34 KiB, allocs estimate: 25.

@benchmark Array{NF,3}(undef, 10, 10, idims...)
BenchmarkTools.Trial: 10000 samples with 199 evaluations.
 Range (min … max):  410.935 ns … 71.406 μs  ┊ GC (min … max):  0.00% … 98.80%
 Time  (median):     594.294 ns              ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.524 μs ±  6.784 μs  ┊ GC (mean ± σ):  59.18% ± 13.03%

  █▁                                                           ▁
  ███▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▁▁▃▁▃▁▁▄▃▃▃▄▄▅▅▄▅▆▅▅▅▆ █
  411 ns        Histogram: log(frequency) by time      50.6 μs <

 Memory estimate: 4.06 KiB, allocs estimate: 1.

Reply:
Of course we could just define a `add!(C, A, B)` function. Maybe that's just the easiest thing to do. 

One of the reasons why the broadcast is slower is that the undef initialiser is slower than for regular arrays. I can't immediately see why that's the case, because the `nonzeros` function that's called by it only takes 1.5 ns, so it's shouldn't matter.

Reply:
> Of course we could just define a add!(C, A, B) function. Maybe that's just the easiest thing to do.

I wouldn't be opposed to that, but if it's just the undef initializer that might be easy to solve? I'll have a look! At the moment it looks like almost all time is spent in the undef initializer?

Reply:
Initially I thought that, but I just tested it and in the `@. C = A + B` case, the undef initializer is not called. So, I assume it is the `copyto!(::LTA, ::Broadcasted)` function. Due to the way broadcast works, there are restrictions on it. The `Broadcasted` object needs to be called either with a linear index (over all dimensions/axes), or with an `CartesianIndex`. Currently we do the latter and also have to count up SH numbers for that. That might be just be enough to make it slower. On GPU it is even slower than that because we have to use the `k2ij` function. 

Again, we could get rid of this problem by not making LTA a subtype of `AbstractArray{T,N}`, but of `AbstractArray{T,N-1}` and pre-computing the `k2ij`. But to be honest, everything works the way we done it now, and for the few cases we use that broadcast in the dynamical core, we can also explicitly define those faster functions.

Reply:
Maybe one GPU just pre-computing the `k2ij` in the current version (`<:AbstractArray{T,N}`) is already enough to make it equal or close in speed.

Reply:
But again, previously we agreed that we don't care so much about the broadcasting speed. If we just use the explicit functions in the dynamical core, we would be fine.

Reply:
How big are the `k2ij`? I thought naively it would need to be a `Int -> (Int, Int)` pair, but you said it was a `BitVector`?

Reply:
I mean, ultimately, we can do both and make both work, I think. It's just additional work to now change everything to `AbstractArray{T,N-1}` that I am not sure we absolutely need to do.

Reply:
`int -> (int, int)` looks like a killer no? At T31 and single precision we'd have

- 1 layer: 560*4 = 2.2KB data but 560*3*8 = 13.4KB for `k2ij` when using `Int64`, so 7x larger = +700%
- 8 layers: +75% larger still
- 8 layers but int32: +37.5% larger
- 8 layers but int16: +18.75% larger
- 2x8 layers and int16: still +9% larger

Reply:
Ultimately, we just have to decide: 
Do we want to use the broadcast in the dynamical core: 
* Yes: Then we should either accerelate it, or change LTA to `AbstractArray{T,N-1}` 
* No: It's fine like it is, and we use those explicit arithmetics (and should add the `add!(C,A,B)` as well)

Reply:
I agree, I think I already had compiled a list of the broadcastable operations that we'd want to define manually (and doing that might be in general smart so that we can change internals without changing the user-facing functionality)